networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -0,0 +1,211 @@
+		outfile.write(reinterpret_cast<char*>(&header.offsetBaseData), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjLists), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetWeights), sizeof(uint64_t));
+	};
+
+	if(!G.isEmpty()) {
+		count prefSum = 0;
+		nodes = G.numberOfNodes();
+		if (nodes < chunks) {
+			chunks = nodes;
+			INFO("reducing chunks to ", chunks, " chunks");
+		}
+		nodesPerChunk = splitNodes(nodes, chunks);
+
+		size_t node_size = 0;
+		count sum = 0;
+		std::vector<count> nodePrefSum;
+		if(!G.isDirected()) {
+			G.forNodes([&](node u) {
+				std::vector<uint64_t> nbr;
+				uint8_t tmp [10];
+				node_size = NetworkitBinaryWriter::encode(u,tmp);
+				G.forNeighborsOf(u,[&](node v){
+					nbr.push_back(v);
+				});
+				sum += nbr.size();
+				nodePrefSum.push_back(sum);
+			});
+		} else {
+			G.forNodes([&](node u) {
+				std::vector<uint64_t> nbr;
+				G.forNeighborsOf(u,[&](node v){
+					nbr.push_back(v);
+				});
+				sum += nbr.size();
+				nodePrefSum.push_back(sum);
+			});
+		}
+
+		/* compute encoded size of arrays and store in vector */
+		size_t prev = 0;
+		std::vector<size_t> encArrSize; //  pref sum of size encoded adj arrays
+		G.forNodes([&](node u) {
+			size_t adj_size = 0;
+			G.forNeighborsOf(u,[&](node v) {
+				uint8_t tmp [10];
+				adj_size += encode(v, tmp);
+			});
+			adj_size += prev;
+			encArrSize.push_back(adj_size);
+			prev = adj_size;
+		});
+		// compute chunks and their offsets
+		std::vector<std::pair<uint64_t, uint64_t>> split; // stores chunk id + first node
+		int ctr = 1;
+		int k = 0;
+		G.forNodes([&](node u) {
+			if (ctr >= nodesPerChunk) {
+				split.push_back(std::make_pair(k,(u+1)));
+				ctr = 1;
+				k++;
+			} else {
+				ctr++;
+			}
+		});
+
+		std::vector<std::pair<uint64_t, uint64_t>> chunksMap;
+		for(uint64_t i = 0; i < chunks; i++) {
+			uint64_t begin;
+			if(!i) {
+				begin = 0;
+			} else {
+				begin = split[i-1].second;
+			}
+
+			uint64_t offset;
+			if(!i) {
+				offset = 0;
+			} else {
+				offset = encArrSize.at(begin-1);
+			}
+			chunksMap.push_back(std::make_pair(begin, offset));
+		}
+
+        // Write header
+		setHeader();
+		writeHeader();
+
+        // Write base data
+		G.forNodes([&] (node u) {
+			outfile.write(reinterpret_cast<char*>(&nodePrefSum.at(u)), sizeof(uint64_t));
+		});
+
+		for (auto &ch : chunksMap) {
+			outfile.write(reinterpret_cast<char*>(&ch.first), sizeof(uint64_t));
+		}
+
+        // Write adjacency data
+		for (auto &ch : chunksMap) {
+			outfile.write(reinterpret_cast<char*>(&ch.second), sizeof(uint64_t));
+		}
+
+		G.forNodes([&](node u) {
+			G.forNeighborsOf(u,[&](node v){
+				uint8_t tmp [10];
+				uint64_t node_size = encode(v, tmp);
+				outfile.write(reinterpret_cast<char*>(tmp), node_size);
+			});
+		});
+
+		INFO("Written graph to ", path);
+	} else {
+		throw std::runtime_error("G is empty.");
```

Comment:
Why is it not possible to store empty graphs? As far as I understood the format, this should be valid.